tests/net/lldp/src/main.c

```diff
@@ -0,0 +1,215 @@
+/* main.c - Application main entry point */
+
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+#include <misc/printk.h>
+#include <linker/sections.h>
+
+#include <ztest.h>
+
+#include <net/ethernet.h>
+#include <net/buf.h>
+#include <net/net_ip.h>
+#include <net/net_if.h>
+
+#define NET_LOG_ENABLED 1
+#include "net_private.h"
+
+#if defined(CONFIG_NET_DEBUG_IF)
+#define DBG(fmt, ...) printk(fmt, ##__VA_ARGS__)
+#else
+#define DBG(fmt, ...)
+#endif
+
+static struct net_if *iface_posix;
+static struct net_if *iface_fake;
+
+static bool test_failed;
+static struct k_sem wait_data;
+
+struct net_if_test {
+	u8_t idx;
+	u8_t mac_addr[sizeof(struct net_eth_addr)];
+	struct net_linkaddr ll_addr;
+};
+
+struct eth_fake_context {
+	struct net_if *iface;
+	u8_t mac_address[6];
+	bool promisc_mode;
+};
+
+static struct eth_fake_context eth_fake_data;
+
+static void eth_fake_iface_init(struct net_if *iface)
+{
+	struct device *dev = net_if_get_device(iface);
+	struct eth_fake_context *ctx = dev->driver_data;
+
+	ctx->iface = iface;
+
+	net_if_set_link_addr(iface, ctx->mac_address,
+			     sizeof(ctx->mac_address),
+			     NET_LINK_ETHERNET);
+
+	ethernet_init(iface);
+
+	net_eth_set_lldpdu(iface);
+}
+
+static int eth_fake_send(struct net_if *iface,
+			 struct net_pkt *pkt)
+{
+	ARG_UNUSED(iface);
+
+	DBG("%s: iface %p pkt %p\n", __func__, iface, pkt);
+
+	struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
+	u16_t type = ntohs(hdr->type);
+
+	DBG("type %x\n", type);
+
+	if (type == NET_ETH_PTYPE_LLDP) {
+		k_sem_give(&wait_data);
+
+		/* TODO: parse packet and set test_failed if needed */
+	}
+
+	net_pkt_unref(pkt);
+
+	return 0;
+}
+
+static enum ethernet_hw_caps eth_fake_get_capabilities(struct device *dev)
+{
+	return ETHERNET_LLDP;
+}
+
+static int eth_fake_set_config(struct device *dev,
+			       enum ethernet_config_type type,
+			       const struct ethernet_config *config)
+{
+	struct eth_fake_context *ctx = dev->driver_data;
+
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_PROMISC_MODE:
```

Comment:
We do not seem to use promiscuous mode flag for anything in this test so the promiscuous mode support be removed